caffe2/onnx/onnx_exporter.cc

```diff
@@ -0,0 +1,585 @@
+OnnxExporter::get_special_operators() const {
+  const static std::unordered_map<std::string, OnnxExporter::SpecialOpConverter>
+      kSpecialOperators = {
+        {"Conv", &OnnxExporter::CreateConvPoolNodes},
+        {"ConvTranspose", &OnnxExporter::CreateConvPoolNodes},
+        {"MaxPool", &OnnxExporter::CreateConvPoolNodes},
+        {"AveragePool", &OnnxExporter::CreateConvPoolNodes},
+        {"FC", &OnnxExporter::CreateGemmNodes},
+        {"Concat", &OnnxExporter::CreateConcatNodes},
+        {"LRN", &OnnxExporter::CreateLrnNodes},
+        {"Reshape", &OnnxExporter::CreateReshapeNodes},
+        {"Slice", &OnnxExporter::CreateSliceNodes},
+        {"ChannelShuffle",  &OnnxExporter::CreateChannelShuffleNodes}
+      };
+  return kSpecialOperators;
+}
+
+void OnnxExporter::CopyCaffe2ArgToOnnxAttr(
+    AttributeProto* attr,
+    const std::string& op_type,
+    const caffe2::Argument& arg) {
+  std::string name;
+  if (get_per_op_renamed_attrs().count(op_type)) {
+    name = caffe2::get_default(
+        get_per_op_renamed_attrs().at(op_type), arg.name(), arg.name());
+  } else {
+    name = caffe2::get_default(get_renamed_attrs(), arg.name(), arg.name());
+  }
+  attr->set_name(name);
+
+  if (arg.has_f()) {
+    attr->set_f(arg.f());
+    attr->set_type(AttributeProto::FLOAT);
+  } else if (arg.has_i()) {
+    attr->set_i(arg.i());
+    attr->set_type(AttributeProto::INT);
+  } else if (arg.has_s()) {
+    attr->set_s(arg.s());
+    attr->set_type(AttributeProto::STRING);
+  } else if (arg.floats_size()) {
+    attr->mutable_floats()->CopyFrom(arg.floats());
+    attr->set_type(AttributeProto::STRINGS);
+  } else if (arg.ints_size()) {
+    attr->mutable_ints()->CopyFrom(arg.ints());
+    attr->set_type(AttributeProto::INTS);
+  } else if (arg.strings_size()) {
+    attr->mutable_strings()->CopyFrom(arg.strings());
+    attr->set_type(AttributeProto::STRINGS);
+  } else {
+    CAFFE_THROW(
+        caffe2::MakeString("Unsupported Caffe2 argument: ", arg.name()));
+  }
+}
+
+bool OnnxExporter::IsBlaskListed(const caffe2::Argument& arg) {
+  const static std::unordered_map<std::string, std::unordered_set<std::string>>
+      kBlackListString = {{"order", {"NCHW"}}};
+  const static std::unordered_map<std::string, std::unordered_set<int64_t>>
+      kBlackListInt = {{"cudnn_exhaustive_search", {0, 1}},
+                       {"use_cudnn", {0, 1}}};
+
+  if (arg.has_i()) {
+    const auto it = kBlackListInt.find(arg.name());
+    if (it != kBlackListInt.end()) {
+      return it->second.count(arg.i());
+    }
+  } else if (arg.has_s()) {
+    const auto it = kBlackListString.find(arg.name());
+    if (it != kBlackListString.end()) {
+      return it->second.count(arg.s());
+    }
+  }
+
+  return false;
+}
+
+ConvertedResult OnnxExporter::Caffe2OpToOnnxNodes(
+    const caffe2::OperatorDef& def,
+    const std::unordered_map<std::string, caffe2::TensorShape>& shapes) {
+  std::string type = def.type();
```

Comment:
You create a copy of the string here, but it doesn't seem necessary. Please use `const string& type = def.type()` instead.

Reply:
Actually, I cannot do this because may be written later.